tree: fix ambiguity with enum overloads

otan opened the pull request:
Resolves #71446 

Release note (sql change): Previously, certain enum builtins or
operators required an explicit enum cast. This has been reduced in some
cases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71653)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
looks very useful! just had small questions

otan reviewed (commented):
bors r=rafiss

otan committed:
tree: fix ambiguity with enum overloads

Release note (sql change): Previously, certain enum builtins or
operators required an explicit enum cast. This has been reduced in some
cases.
